File: monitoring/middleware/performance-logger.js
/**
 * Performance Logging Middleware for Express.js
 * Tracks response times, status codes, and request details
 */

const fs = require('fs');
const path = require('path');

class PerformanceLogger {
  constructor(options = {}) {
    this.logDir = options.logDir || path.join(__dirname, '../logs');
    this.logFile = path.join(this.logDir, 'performance.log');
    this.metricsFile = path.join(this.logDir, 'metrics.json');
    this.enableConsole = options.enableConsole !== false;
    
    // Create logs directory if it doesn't exist
    if (!fs.existsSync(this.logDir)) {
      fs.mkdirSync(this.logDir, { recursive: true });
    }
    
    // Initialize metrics storage
    this.metrics = {
      requests: [],
      summary: {
        totalRequests: 0,
        totalErrors: 0,
        averageResponseTime: 0,
        startTime: new Date().toISOString()
      }
    };
    
    this.loadMetrics();
  }

  loadMetrics() {
    try {
      if (fs.existsSync(this.metricsFile)) {
        const data = fs.readFileSync(this.metricsFile, 'utf8');
        this.metrics = JSON.parse(data);
      }
    } catch (error) {
      console.error('Error loading metrics:', error.message);
    }
  }

  saveMetrics() {
    try {
      fs.writeFileSync(this.metricsFile, JSON.stringify(this.metrics, null, 2));
    } catch (error) {
      console.error('Error saving metrics:', error.message);
    }
  }

  middleware() {
    return (req, res, next) => {
      const startTime = Date.now();
      const startDate = new Date();

      // Capture original end function
      const originalEnd = res.end;

      res.end = (...args) => {
        const duration = Date.now() - startTime;
        
        const logEntry = {
          timestamp: startDate.toISOString(),
          method: req.method,
          path: req.path,
          url: req.originalUrl || req.url,
          statusCode: res.statusCode,
          duration: duration,
          ip: req.ip || req.connection.remoteAddress,
          userAgent: req.get('user-agent') || 'unknown'
        };

        // Log to file
        this.logToFile(logEntry);
        
        // Update metrics
        this.updateMetrics(logEntry);
        
        // Console output
        if (this.enableConsole) {
          const color = this.getColorForStatus(res.statusCode);
          console.log(
            `${logEntry.timestamp} | ${color}${res.statusCode}\x1b[0m | ` +
            `${duration}ms | ${req.method} ${req.path}`
          );
        }

        // Call original end
        originalEnd.apply(res, args);
      };

      next();
    };
  }

  logToFile(entry) {
    const logLine = JSON.stringify(entry) + '\n';
    fs.appendFileSync(this.logFile, logLine);
  }

  updateMetrics(entry) {
    this.metrics.requests.push(entry);
    this.metrics.summary.totalRequests++;
    
    if (entry.statusCode >= 400) {
      this.metrics.summary.totalErrors++;
    }

    // Calculate average response time
    const totalTime = this.metrics.requests.reduce((sum, req) => sum + req.duration, 0);
    this.metrics.summary.averageResponseTime = Math.round(totalTime / this.metrics.requests.length);

    // Keep only last 10000 requests in memory
    if (this.metrics.requests.length > 10000) {
      this.metrics.requests = this.metrics.requests.slice(-10000);
    }

    // Save metrics every 100 requests
    if (this.metrics.summary.totalRequests % 100 === 0) {
      this.saveMetrics();
    }
  }

  getColorForStatus(statusCode) {
    if (statusCode >= 500) return '\x1b[31m'; // Red
    if (statusCode >= 400) return '\x1b[33m'; // Yellow
    if (statusCode >= 300) return '\x1b[36m'; // Cyan
    if (statusCode >= 200) return '\x1b[32m'; // Green
    return '\x1b[0m'; // Reset
  }

  getStats() {
    return {
      ...this.metrics.summary,
      errorRate: ((this.metrics.summary.totalErrors / this.metrics.summary.totalRequests) * 100).toFixed(2) + '%',
      requestsInMemory: this.metrics.requests.length
    };
  }
}

module.exports = PerformanceLogger;

File: monitoring/middleware/performance_logger.py
"""
Performance Logging Middleware for Django
Tracks response times, status codes, and request details
"""

import json
import time
import logging
from datetime import datetime
from pathlib import Path
from django.utils.deprecation import MiddlewareMixin

# Configure logging
log_dir = Path(__file__).parent.parent / 'logs'
log_dir.mkdir(exist_ok=True)

logging.basicConfig(
    filename=log_dir / 'performance.log',
    level=logging.INFO,
    format='%(message)s'
)

logger = logging.getLogger('performance')


class PerformanceLoggingMiddleware(MiddlewareMixin):
    """
    Django middleware to log request/response performance metrics
    """
    
    def __init__(self, get_response):
        self.get_response = get_response
        self.metrics_file = log_dir / 'metrics.json'
        self.metrics = self.load_metrics()
    
    def load_metrics(self):
        """Load existing metrics from file"""
        try:
            if self.metrics_file.exists():
                with open(self.metrics_file, 'r') as f:
                    return json.load(f)
        except Exception as e:
            logger.error(f"Error loading metrics: {e}")
        
        return {
            'requests': [],
            'summary': {
                'total_requests': 0,
                'total_errors': 0,
                'average_response_time': 0,
                'start_time': datetime.now().isoformat()
            }
        }
    
    def save_metrics(self):
        """Save metrics to file"""
        try:
            with open(self.metrics_file, 'w') as f:
                json.dump(self.metrics, f, indent=2)
        except Exception as e:
            logger.error(f"Error saving metrics: {e}")
    
    def process_request(self, request):
        """Mark request start time"""
        request._start_time = time.time()
        return None
    
    def process_response(self, request, response):
        """Log response metrics"""
        if hasattr(request, '_start_time'):
            duration = (time.time() - request._start_time) * 1000  # Convert to ms
            
            log_entry = {
                'timestamp': datetime.now().isoformat(),
                'method': request.method,
                'path': request.path,
                'status_code': response.status_code,
                'duration': round(duration, 2),
                'ip': self.get_client_ip(request),
                'user_agent': request.META.get('HTTP_USER_AGENT', 'unknown')
            }
            
            # Log to file
            logger.info(json.dumps(log_entry))
            
            # Update metrics
            self.update_metrics(log_entry)
            
            # Console output
            color = self.get_color_for_status(response.status_code)
            reset = '\033[0m'
            print(
                f"{log_entry['timestamp']} | {color}{response.status_code}{reset} | "
                f"{duration:.0f}ms | {request.method} {request.path}"
            )
        
        return response
    
    def get_client_ip(self, request):
        """Extract client IP address"""
        x_forwarded_for = request.META.get('HTTP_X_FORWARDED_FOR')
        if x_forwarded_for:
            ip = x_forwarded_for.split(',')[0]
        else:
            ip = request.META.get('REMOTE_ADDR')
        return ip
    
    def update_metrics(self, entry):
        """Update metrics summary"""
        self.metrics['requests'].append(entry)
        self.metrics['summary']['total_requests'] += 1
        
        if entry['status_code'] >= 400:
            self.metrics['summary']['total_errors'] += 1
        
        # Calculate average response time
        total_time = sum(req['duration'] for req in self.metrics['requests'])
        self.metrics['summary']['average_response_time'] = round(
            total_time / len(self.metrics['requests']), 2
        )
        
        # Keep only last 10000 requests in memory
        if len(self.metrics['requests']) > 10000:
            self.metrics['requests'] = self.metrics['requests'][-10000:]
        
        # Save metrics every 100 requests
        if self.metrics['summary']['total_requests'] % 100 == 0:
            self.save_metrics()
    
    def get_color_for_status(self, status_code):
        """Return ANSI color code for status"""
        if status_code >= 500:
            return '\033[31m'  # Red
        if status_code >= 400:
            return '\033[33m'  # Yellow
        if status_code >= 300:
            return '\033[36m'  # Cyan
        if status_code >= 200:
            return '\033[32m'  # Green
        return '\033[0m'  # Reset
